### PR TITLE
Ticket #33676 : Enhancement stop waiting a set time for microdescriptors

### PR DESCRIPTION
**Pull request by ANURADHAJHA99**
The following PR is to run  the Travis CI tests and make further changes according to the provided instructions in the ticket. I have added all the commits since all the changes are inter-related to each other.

### CODE REVIEW

**Comment by ANURADHAJHA99** [comment]
hey @teor2345 , can you please guide me regarding why the build failed? My further changes are based on it. Thank you very much.

**Comment by teor2345** [comment]
Hi, I replied on the ticket:

>  I think we need to fix this issue in #33675, before the maint-0.3.5 build will pass:
> > if we are not getting the right value for ed25519_id, move _setEd25519Id() into LocalNodeController, so we can use it during wait_for_bootstrap()

https://trac.torproject.org/projects/tor/ticket/33676#comment:2

I wonder if your code is passing locally on master and maint-0.3.5:

> Your code in #33675 needs to pass your local "make test-network-all", on master and maint-0.3.5, before you start this ticket.

https://trac.torproject.org/projects/tor/ticket/33676#comment:description

Did you test your PR locally on maint-0.3.5 and master?

Did you add logs for the ed25519_key in the pattern?

What do the logs say?

**Comment by ANURADHAJHA99** [comment]
Hey @teor2345 , yes it was passing for the main-0.3.5 and master both.

**Comment by ANURADHAJHA99** [comment]
I checked in cached-microdescs.new for ed25519_id and command  

> cat chutney/net/nodes/000a/keys/ed25519_master_id_public_key | tail -c 32 | base64 | head -c -2
and both the results are same. Do you think the search value is yielding wrong?

**Comment by teor2345** [comment]
> Did you add logs for the ed25519_key in the pattern?
> 
> What do the logs say?

Please see what I wrote in #33675:

> > if md is True, print the values of nickname and ed25519_key in getNodeDirInfoStatusPattern()
> > if we are not getting the right value for ed25519_id, move _setEd25519Id() into LocalNodeController, so we can use it during wait_for_bootstrap() 

https://trac.torproject.org/projects/tor/ticket/33675#comment:25

**Comment by ANURADHAJHA99** [comment]
Hey @teor2345 , because of internet issue, i will probably miss your advise on IRC so mentioning the solution here but I solved this problem, i was calling the function wrong, my bad. I have fixed it and by printing the nickname and ed25519_key, I am finally getting the right value. I have attached the logs, I will go ahead with further changes for stopping the wait time after the travis CI run succeeds. Thank you very much for your help
[single-onion-v23-ipv6-md.log](https://github.com/torproject/chutney/files/4458003/single-onion-v23-ipv6-md.log)

**Comment by ANURADHAJHA99** [comment]
Hi @teor2345 , the log for the following issue is still failing, though I don't necessarily see the problem, what could go wrong? I can only make further changes if all my tests are passing. Thank you very much.

File: lib/chutney/TorNet.py [hunk 1]
```diff
@@ -52,7 +53,6 @@ def getenv_type(env_var, default, type_, type_name=None):
     """
        Return the value of the environment variable 'envar' as type_,
        or 'default' if no such variable exists.
-
```
Comment (teor2345) [inline lib/chutney/TorNet.py:55]: Why did you delete all these newlines?
What editor are you using?

We try to make our diffs as small as possible, so we don't have conflicts.
So please don't change whitespace in unrelated code.

File: lib/chutney/TorNet.py [hunk 2]
```diff
@@ -667,6 +650,7 @@ def preConfig(self, net):
             self._genAuthorityKey()
         if self._env['relay']:
             self._genRouterKey()
+            #self._setEd25519Id()
```
Comment (teor2345) [inline lib/chutney/TorNet.py:653]: You can delete this line, it will be in the git history.

File: lib/chutney/TorNet.py [hunk 3]
```diff
@@ -763,7 +746,7 @@ def _genRouterKey(self):
                   .format(" ".join(cmdline), stdouterr))
             sys.exit(1)
         self._env['fingerprint'] = fingerprint
-
+
```
Comment (teor2345) [inline lib/chutney/TorNet.py:749]: Please don't change this whitespace.

File: lib/chutney/TorNet.py [hunk 4]
```diff
@@ -863,6 +871,15 @@ def getBridge(self):
         except KeyError:
             return 0
 
+    def getEd25519Id(self):
+        """Return the value of ed25519 key"""
+        try:
+            self._setEd25519Id()
+            return self._env['ed25519_id']
+        except KeyError:
+
```
Comment (teor2345) [inline lib/chutney/TorNet.py:880]: You don't need this empty line here.
It's harder to read the code with this gap.

File: lib/chutney/TorNet.py [hunk 5]
```diff
@@ -863,6 +871,15 @@ def getBridge(self):
         except KeyError:
             return 0
 
+    def getEd25519Id(self):
+        """Return the value of ed25519 key"""
+        try:
+            self._setEd25519Id()
```
Comment (teor2345) [inline lib/chutney/TorNet.py:877]: You're calling `_setEd25519Id()` every time you want to get `_env['ed25519_id']`, but you should only need to call it once per node.

The code could work this way, but it's not very efficient.

Can you call `_setEd25519Id()` from `__init__()` instead?

File: lib/chutney/TorNet.py [hunk 6]
```diff
@@ -1294,13 +1304,11 @@ def getNodePublishedDirInfoPaths(self):
 
     def getNodeDirInfoStatusPattern(self, dir_format):
         """Returns a regular expression pattern for finding this node's entry
-           in a dir_format file.
-
-           The microdesc format is not yet implemented, because it requires
-           the ed25519 key. (Or RSA block matching, which is hard.)
+           in a dir_format file and returning None if nickname or ed25519_id key not found.
```
Comment (teor2345) [inline lib/chutney/TorNet.py:1307]: A few fixes:
* nickname is never checked for None, because it's a required variable.  So it never causes this function to return None.
* there are two different ways that this function can return None
* we want to explain why the function returns None, and what that means. We don't need to repeat exactly what the code says.
```suggestion
           in a dir_format file. Returns None if the requested pattern is not available.
```

File: lib/chutney/TorNet.py [hunk 7]
```diff
@@ -1323,16 +1331,17 @@ def getNodeDirInfoStatusPattern(self, dir_format):
...
             return r'^router ' + nickname + " "
         elif md:
-            # Not yet implemented, see #33428
-            # r'^id ed25519 " + ed25519_identity (end of line)
-            # needs ed25519-identity from #30642
-            # or the existing keys/ed25519_master_id_public_key
-            return None
-
+            print(nickname)
+            print(ed25519_key)
```
Comment (teor2345) [inline lib/chutney/TorNet.py:1335]: You don't need these print statements any more.

File: lib/chutney/TorNet.py [hunk 8]
```diff
@@ -2467,4 +2452,4 @@ def main():
     return 0
 
 if __name__ == '__main__':
-    sys.exit(main())
+    sys.exit(main())
```
Comment (teor2345) [inline lib/chutney/TorNet.py:2455]: Please don't delete this newline.

You might find it easier to do this work with a better editor.

File: lib/chutney/TorNet.py (hunk 5 above)
Comment (ANURADHAJHA99) [reply lib/chutney/TorNet.py:877]: hi @teor2345 , can you elaborate this part for me a bit, should I just call the function inside  _init_ ? I am getting errors in doing so.

Comment (teor2345) [reply lib/chutney/TorNet.py:877]: What errors are you getting?

Comment (ANURADHAJHA99) [reply lib/chutney/TorNet.py:877]: FileNotFoundError: [Errno 2] No such file or directory: '/home/anuradha/chutney/net/nodes/002r/keys/ed25519_master_id_public_key'   I did not get the same error while calling the function inside getEd25519Id().

Comment (teor2345) [reply lib/chutney/TorNet.py:877]: That looks like the right path.

Can you push the code you're actually using?
It's hard to give advice when I can't see your code.

File: lib/chutney/TorNet.py [hunk 9]
```diff
@@ -176,10 +167,13 @@ def get_absolute_nodes_path():
        Returns the absolute path of the "nodes" symlink that points to the
        "nodes*" directory that chutney should use to store the current
        network's torrcs and tor runtime data.
+<<<<<<< HEAD
+=======
 
        This path is also used as a prefix for the unique nodes directory
        names.
 
+>>>>>>> 6282f471e432efa2aadf1046ea6e0fbaca17b434
```
Comment (teor2345) [inline lib/chutney/TorNet.py:176]: Please remove the merge conflict markers in your code.

File: lib/chutney/TorNet.py [hunk 10]
```diff
@@ -849,11 +834,39 @@ def _getBridgeLines(self):
...
+                EXPECTED_ED25519_BASE64_KEY_SIZE = 43
+                CURRENT_ED25519_BASE64_KEY_SIZE = len(ed25519_id)
+                if CURRENT_ED25519_BASE64_KEY_SIZE != EXPECTED_ED25519_BASE64_KEY_SIZE:
+                    raise ValueError("The current length of the key is {}, which is not matching the expected length of {}".format(CURRENT_ED25519_BASE64_KEY_SIZE, EXPECTED_ED25519_BASE64_KEY_SIZE))
+                else:
+                    self._env['ed25519_id'] = ed25519_id
+    
     def __init__(self, env):
-        NodeController.__init__(self, env)
+        NodeController.__init__(self,env)
```
Comment (teor2345) [inline lib/chutney/TorNet.py:865]: Please don't make unnecessary edits.

File: lib/chutney/TorNet.py (hunk 5 above)
Comment (teor2345) [reply lib/chutney/TorNet.py]: Where do you get the error in the logs?

Please link to the full logs in CI.

Comment (ANURADHAJHA99) [reply lib/chutney/TorNet.py]: https://travis-ci.org/github/torproject/chutney/jobs/675703309 on line 912.

**Comment by teor2345** [comment]
Squashed and fixed in #72.
